pkg/client/typed_client.go

```diff
@@ -159,8 +159,8 @@ func (c *typedClient) List(ctx context.Context, obj ObjectList, opts ...ListOpti
 		Into(obj)
 }
 
-// UpdateStatus used by StatusWriter to write status.
-func (c *typedClient) UpdateStatus(ctx context.Context, obj Object, opts ...UpdateOption) error {
+// UpdateSubResource used by SubResourceWriter to write status.
+func (c *typedClient) UpdateSubResource(ctx context.Context, obj Object, subResource string, opts ...UpdateOption) error {
```

Comment:
IMHO, one `obj Object` in the parameters is not enough, because not all subresources (even CRUD subresources) take the original object as its request body. For subresources like `Status` and `Finalize`, yes. For subresources like `Scale`, no, which actually accept a `autoscaling.Scale` as the request body.

Reply:
Yeah, we definitely need that, apart from Scale there is also `authenticationv1.TokenRequest` in the case of serviceaccount tokens...Maybe we can have a `StatusUpdateOption` that has a `Body Object` field on top of the other `UpdateOptions`?

Reply:
SGTM, or we can have two objects in the UpdateSubResource parameters like
`(ctx context.Context, obj Object, subObj Object, subResource string, opts ...UpdateOption)`

If the `subObj` is not nil, we should take it as the update body, otherwise we will use the `obj`.

Reply:
Due to the fact that it is not always mandatory I'd prefer to make it part of the opts and we can then provide a convenient wrapper, so that you can do something like:

`c.UpdateSubresource(ctx, myObj, "my-subresource", client.WithSubResourceBody(mySubResource))`